xocto/ranges.py

```diff
@@ -1151,3 +1165,19 @@ def iterate_over_months(
 
         yield FiniteDatetimeRange(start_at, this_end)
         start_at = next_start
+
+
+# Subscripted generics may not be used with isinstance directly.
+# TODO: A TypeGuard would be nicer, once we drop Python 3.9.
+def _is_datetime_range(value: Any) -> bool:
+    if value.start is not None and not isinstance(value.start, datetime.datetime):
+        return False
+    if value.end is not None and not isinstance(value.end, datetime.datetime):
+        return False
+    return True
+
+
+# Subscripted generics may not be used with isinstance directly.
+# TODO: A TypeGuard would be nicer, once we drop Python 3.9.
+def _is_half_finite_datetime_range(value: Any) -> bool:
```

Comment:
I moved this function from `xocto/fields/postgres/ranges` to here, since I need the same function here and this seems like the better home for it. It means we're calling a private function of `xocto/ranges` in `xocto/fields/postgres/ranges`, but I'm not aware of a better way to do this. I don't want to make this function public (`def is_half_finite_datetime_range`), since then it would also be part of the external public API. Ideally I'd want some kind of "public within this package" access, but python doesn't have anything like that... I think this is probably fine as it is 🤷

Reply:
Probably what can be done is to make a “private package” and make the function itself public but that would involve a major restructuring so not actually suggesting now